File: openapi/linter/customrules/types/src/rule.ts
import type { Context } from './context';
import type { RuleConfig } from './config';
import type { DocumentInfo } from './document';
import type { Severity } from './severity';
import type { ValidationError } from './validation';

/**
 * RuleRunner is the interface that custom rules must implement.
 */
export interface RuleRunner {
  /**
   * Unique identifier for this rule (e.g., "custom-require-summary").
   * Should be prefixed with "custom-" to avoid conflicts with built-in rules.
   */
  id(): string;

  /**
   * Rule category (e.g., "style", "semantic", "security", "schemas").
   */
  category(): string;

  /**
   * Detailed description of what the rule checks.
   */
  description(): string;

  /**
   * Short summary of what the rule checks.
   */
  summary(): string;

  /**
   * Optional URL to documentation for this rule.
   */
  link(): string;

  /**
   * Default severity level for this rule.
   */
  defaultSeverity(): Severity;

  /**
   * OpenAPI versions this rule applies to, or null for all versions.
   * Example: ["3.0", "3.1"] to only run on OpenAPI 3.x documents.
   */
  versions(): string[] | null;

  /**
   * Execute the rule against the document.
   *
   * @param ctx - Context for cancellation checking
   * @param docInfo - Document information including the parsed document and index
   * @param config - Rule configuration including severity overrides
   * @returns Array of validation errors found
   */
  run(ctx: Context, docInfo: DocumentInfo, config: RuleConfig): ValidationError[];
}

/**
 * Base class for implementing custom rules.
 * Provides default implementations for optional methods.
 *
 * @example
 * ```typescript
 * class RequireOperationSummary extends Rule {
 *   id(): string { return 'custom-require-operation-summary'; }
 *   category(): string { return 'style'; }
 *   description(): string { return 'All operations must have a summary.'; }
 *   summary(): string { return 'Operations must have summary'; }
 *
 *   run(ctx: Context, docInfo: DocumentInfo, config: RuleConfig) {
 *     const errors: ValidationError[] = [];
 *     for (const opNode of docInfo.index.operations) {
 *       const op = opNode.node;
 *       if (!op.getSummary()) {
 *         errors.push(createValidationError(
 *           config.getSeverity(this.defaultSeverity()),
 *           this.id(),
 *           'Operation is missing a summary',
 *           op.getRootNode()
 *         ));
 *       }
 *     }
 *     return errors;
 *   }
 * }
 *
 * registerRule(new RequireOperationSummary());
 * ```
 */
export abstract class Rule implements RuleRunner {
  abstract id(): string;
  abstract category(): string;
  abstract description(): string;
  abstract summary(): string;

  /** Default implementation returns empty string (no documentation link). */
  link(): string {
    return '';
  }

  /** Default implementation returns 'warning'. */
  defaultSeverity(): Severity {
    return 'warning';
  }

  /** Default implementation returns null (applies to all versions). */
  versions(): string[] | null {
    return null;
  }

  abstract run(ctx: Context, docInfo: DocumentInfo, config: RuleConfig): ValidationError[];
}

/**
 * Register a rule with the linter.
 * Call this at the end of your rule file to make it available.
 *
 * @param rule - The rule instance to register
 *
 * @example
 * ```typescript
 * registerRule(new MyCustomRule());
 * ```
 */
declare function registerRule(rule: RuleRunner): void;

export { registerRule };

File: linter/rule.go
package linter

import (
	"context"

	"github.com/speakeasy-api/openapi/validation"
)

// Rule represents a single linting rule
type Rule interface {
	// ID returns the unique identifier for this rule (e.g., "style-path-params")
	ID() string

	// Category returns the rule category (e.g., "style", "validation", "security")
	Category() string

	// Description returns a human-readable description of what the rule checks
	Description() string

	// Summary returns a short summary of what the rule checks
	Summary() string

	// Link returns an optional URL to documentation for this rule
	Link() string

	// DefaultSeverity returns the default severity level for this rule
	DefaultSeverity() validation.Severity

	// Versions returns the spec versions this rule applies to (nil = all versions).
	// Supports exact versions ("3.1") and prefix matching ("3.0" matches "3.0.x").
	Versions() []string
}

// RuleRunner is the interface rules must implement to execute their logic
// This is separate from Rule to allow different runner types for different specs
type RuleRunner[T any] interface {
	Rule

	// Run executes the rule against the provided document
	// DocumentInfo provides both the document and its location for resolving external references
	// Returns any issues found as validation errors
	Run(ctx context.Context, docInfo *DocumentInfo[T], config *RuleConfig) []error
}

// DocumentedRule provides extended documentation for a rule
type DocumentedRule interface {
	Rule

	// GoodExample returns YAML showing correct usage
	GoodExample() string

	// BadExample returns YAML showing incorrect usage
	BadExample() string

	// Rationale explains why this rule exists
	Rationale() string

	// FixAvailable returns true if the rule provides auto-fix suggestions
	FixAvailable() bool
}

// ConfigurableRule indicates a rule has configurable options
type ConfigurableRule interface {
	Rule

	// ConfigSchema returns JSON Schema for rule-specific options
	ConfigSchema() map[string]any

	// ConfigDefaults returns default values for options
	ConfigDefaults() map[string]any
}
